### PR TITLE
Update headers upon a 304; fixes #97

### LINKED ISSUE #97: 206/304 handling broken?
@mnot, at some point you mentioned the handling of 304 was broken. Presumably that also applies to 206, which is not really handled at all at the moment. Could you elaborate on that or point to where we discussed this?

### PR DESCRIPTION
**Pull request by mnot**

### CODE REVIEW

**Comment by annevk** [comment]
I don't think this was the main problem. The problem was that 304/206 were handled at an earlier stage than the current draft suggests. And this cannot fix that issue as it does not handle 206. Or am I missing something?

**Comment by mnot** [comment]
I agree that 206 needs to be handled. Not sure what you mean by "earlier stage" though -- the underlying issue AFAIK was that response headers weren't updated by the 304. Please explain?

**Comment by mnot** [comment]
WRT 206 - a decision similar to the one made for validation needs to be made -- what to do when there's a Range header on the incoming request. Is it worth creating a whole new mode for "don't combine partial responses with what you have cached, pass them through"?

**Comment by annevk** [comment]
So reading that thread linked from the issue. One of the problems is that where 304 is handled now in Fetch, CORS has already happened. So the 304 would be rejected if it lacked CORS headers. Which per that thread is wrong. We should first get the response from the cache, update things all around, and then do a CORS check.

**Comment by mnot** [comment]
Ah, right. Will dig more.

**Comment by mnot** [comment]
I think I need to move the 304 case (and presumably 206 when it materialises) into HTTP-network-or-cache fetch. Make sense?

**Comment by annevk** [comment]
Yeah, I think that's what we concluded.

**Comment by mnot** [comment]
I feel like 

> If request's cache mode is "default" and request's header list contains a header named `If-Modified-Since`, `If-None-Match`, `If-Unmodified-Since`, `If-Match`, or `If-Range`, set request's cache mode to "no-store".

should go in there too. Any objection?

**Comment by annevk** [comment]
Sounds good.

**Comment by mnot** [comment]
OK, take a look RE: 304. 

What do you want to do with 206?

**Comment by annevk** [comment]
It seems the `<!-- XXX xref "HTTP cache" -->` comments can be removed. And for 206 I guess we want something similar. https://github.com/whatwg/fetch/issues/38#issuecomment-125277789 has some comments on how Gecko only supports a single partial entry, but I think the specification could support multiple if that doesn't complicate things too much.

**Comment by mnot** [comment]
I'd like to do 206 as a separate discussion, because it could be pretty intrusive. Talking to @jakearchibald  last night, it's not clear how capable we want to make it. 

Can we push that out of #97 into a separate issue?

**Comment by annevk** [comment]
Separating that out seems fine. Comments still need to be removed, maybe?

File: Overview.src.html [hunk 1]
```diff
@@ -2481,6 +2434,18 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
  <var title>httpRequest</var>'s
  <span title=concept-request-header-list>header list</span>.
 
+ <li><p>If <var title>httpRequest</var>'s
```
Comment (annevk) [inline Overview.src.html]: s/`<var title>`/`<var>`/ for changed `<var>`s.

File: Overview.src.html [hunk 2]
```diff
@@ -2602,6 +2567,52 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
    <var title>httpRequest</var> with the <i title>credentials flag</i> set if set.
   </ol>
 
+ <li>
+   <p>If <var title>response</var>'s <span title=status>status</span> is <code>304</code>, run
```
Comment (annevk) [inline Overview.src.html]: Single-space indentation.

File: Overview.src.html [hunk 3]
```diff
@@ -2602,6 +2567,52 @@ <h4 id="should-response-to-request-be-blocked-due-to-nosniff?"><dfn title="shoul
...
+      <span title=concept-response-status>status</span> which is most likely <code title>200</code>
+      now.
+
+      <p>Set <var title>response</var>'s
+      <span title=concept-response-cache-state>cache state</span> to
+      "<code title>validated</code>".
+    </ol>
+  </ol>
+
+
```
Comment (annevk) [inline Overview.src.html]: One newline less here.

File: Overview.src.html (hunk 1 above)
Comment (mnot) [inline Overview.src.html]: what is a changed <var>?

Comment (annevk) [inline Overview.src.html]: Basically any that is on a green line in this diff view thingie.

**Comment by annevk** [comment]
I ended up having to fix the hyperlinks a bit since you had two closing tags for them. But other than that this was mostly fine. Thank you!

**Pull request closed without merging**